Create tarballs for packages and upload as artifact for PRs (#225)
theforeman/foreman-js: CONTRIBUTING.md

5. Open [this project on GitHub](https://github.com/theforeman/foreman-js), then click “Compare & pull request”.

## Testing Pull Requests

> 1. Open pull request in Github and go to the tasks tab.

Reviewer: ```suggestion
1. Open the pull request in Github and go to the __Checks__ tab.
```

> 2. Expand `Build tarballs for testing` jobs and open the relevant action. For example, the job for the test package would be `build-tarball (test)`.

Reviewer: The artifact seems to be available no matter which action I select in the nav bar.
```suggestion
2. Expand `Build tarballs for testing` and open any of the associated actions. For example, the job for the test package would be `build-tarball (test)`.
```

3. Expand the artifacts dropdown on the top right and download the appropiate artifact. For example, the test package will be titled `foreman-js-test-pr-###.tgz`.
4. Unzip and copy the unpacked tarball to your development box or the appropiate location:
```bash
# download `foreman-js-test-pr-123.tgz.zip` from `artifacts` dropdown on top right of github action
cd /path/to/download_dir && unzip foreman-js-test-pr-123.tgz.zip
scp theforeman-test-4.15.1.tgz vagrant@centos7-katello-devel-stable.example.com:~
```

>   - Side note: ideally we could use the tarball URL directly but looks like there is a permission issue that will make it hard to use with npm: https://github.com/actions/upload-artifact/issues/51

Reviewer: Would it work to have a shell script download the tarball with `curl`?
If so, it wouldn't be too hard to have a script that automates steps 4 through 8-- you could just pass in the url.

Author: From the issue, it sounds like `curl -v -H "Authorization: token $GITHUB_TOKEN" $TARBALL_URL` would work

We could try out the curl command and give it as an option if it works as expected. I would just want to make sure the permissions don't block it for non-maintainers (or we make that distinction clear in the docs)

5. Update package.json in foreman or plugin to use local tarball:
```
$ cat package.json | grep theforeman-test
    "@theforeman/test": "/home/vagrant/theforeman-test-4.15.1.tgz",
```

> 6. `rm -rf node_modules && rm package-lock.json` in foreman or plugin.

Reviewer: This makes it about as much work as the other `npm i; npm run build; npm run link` workflow.  I wonder if there's any way to tell npm to be more selective or just to replace a single package..

Author: The benefits to me are:
A) Module resolution happens naturally, so any executables provided by a package function as expected and interactions with other packages outside of @theforeman ones resolve as expected.
B) No longer having to worry about symlinked files being modified. I've too often symlinked to foreman, forgot about it, and then ran `npm i` and overwrote my foreman-js folder.
C) It's easier for those outside of those not knowledgeable about this repo to review PRs

I get that it takes some time to reinstall everything, but it's unattended and we can be sure that the packages are resolving as they will once the new @theforeman package is published. To me it's not about saving time, more having a way to test that will function the same way as the published package will.

Of course I am fine with keeping `npm run link`, I can understand if people are used to that workflow or prefer it.
